### PR TITLE
Example on multimodal entailment

### PR DESCRIPTION
**Pull request by sayakpaul**
Textual entailment is a well-studied problem and is also a part of the GLUE benchmark. It's also highly useful for curating and moderating content on social media platforms. As we know social media content contains different data modalities -- texts, images, videos, audios, etc. Because of the nature of interaction that takes place on these platforms, it may be difficult to just use a single modality to learn the entailment task. This is beautifully presented in this ACL tutorial: https://multimodal-entailment.github.io/. 

The tutorial mentioned above introduces a multimodal entailment dataset consisting of tweets and corresponding images labeled as entailment, contradictory, and no_entailment. 

This example presents a baseline implementation of a multimodal model on the said dataset for encouraging further research in the field. [Here](https://colab.research.google.com/github/sayakpaul/Multimodal-Entailment-Baseline/blob/main/multimodal_entailment.ipynb) is a Colab Notebook version.

### CODE REVIEW

File: examples/nlp/multimodal_entailment.py [hunk 1]
```diff
@@ -0,0 +1,565 @@
...
+    vision_encoder = create_vision_encoder(
+        num_projection_layers, projection_dims, dropout_rate, vision_trainable
+    )
+    text_encoder = create_text_encoder(
+        num_projection_layers, projection_dims, dropout_rate, text_trainable
+    )
+
+    # Fetch the embedding projections.
+    vision_projections = vision_encoder([image_1, image_2])
+    text_projections = text_encoder(text_inputs)
```
Comment (sayakpaul) [inline examples/nlp/multimodal_entailment.py:507]: @fchollet I want to incorporate a modality dropout trick here to account for the following situation. What if one of the inputs or a pair of the input is not present during inference? How do we make our model perform well under those situations? 

Apparently, modality dropout can be a useful recipe that can help. I am thinking of doing something like the following:

```python
# Fetch the embedding projections.
vision_projections = vision_encoder([image_1, image_2])
vision_projections = keras.layers.Dropout(0.2)(vision_projections)
text_projections = text_encoder(text_inputs)
text_projections = keras.layers.Dropout(0.2)(text_projections)
```

Does this make sense? Or should we introduce `Dropout` at the inputs themselves? I had tried that but it's failing at the text inputs.

Comment (sayakpaul) [reply examples/nlp/multimodal_entailment.py:499]: I had another idea on incorporating cross-attention here:

```python
vision_projections = vision_encoder([image_1, image_2])
text_projections = text_encoder(text_inputs)

# Cross-attention (Luong-style).
query_value_attention_seq = keras.layers.Attention(use_scale=True, dropout=0.2)(
    [vision_projections, text_projections]
)

# Concatenate the projections and pass through the classification layer.
concatenated = keras.layers.Concatenate()([vision_projections, text_projections])
contextual = keras.layers.Concatenate()([concatenated, query_value_attention_seq])
outputs = keras.layers.Dense(3, activation="softmax")(contextual)
```

Would love to get your thoughts on both of the above ideas.

**Comment by sayakpaul** [comment]
@rchao @fchollet just wanted to gently ping about this PR.

**Comment by rchao** [comment]
@sayakpaul sorry for the delay. It's been quite busy and we're a bit short-staffed so thanks for your patience. We're try to get to this asap.

File: examples/nlp/multimodal_entailment.py (hunk 1 above)
Comment (fchollet) [reply examples/nlp/multimodal_entailment.py:499]: Both of these sound fine to me as a way to add regularization. Any particular concern you had?

File: examples/nlp/multimodal_entailment.py [hunk 2]
```diff
@@ -0,0 +1,577 @@
...
+"""shell
+!pip install -q tensorflow_text
+"""
+
+"""
+## Collect the dataset
+
+The original dataset is available
+[here](https://github.com/google-research-datasets/recognizing-multimodal-entailment).
+However, we will be using a better prepared version of the dataset. Thanks to
```
Comment (fchollet) [inline examples/nlp/multimodal_entailment.py:42]: "we will be using a better prepared version" -> please detail how it was modified from the original and why.

File: examples/nlp/multimodal_entailment.py [hunk 3]
```diff
@@ -0,0 +1,577 @@
...
+[Nilabhra Roy Chowdhury](https://de.linkedin.com/in/nilabhraroychowdhury) who worked on
+the preparation. 
+"""
+
+"""shell
+!wget -q https://github.com/sayakpaul/Multimodal-Entailment-Baseline/releases/download/v1.0.0/tweet_images.tar.gz
```
Comment (fchollet) [inline examples/nlp/multimodal_entailment.py:48]: Prefer using Python-based download and extraction, e.g. `get_file()`, so that the script can be run in a standalone way

File: examples/nlp/multimodal_entailment.py [hunk 4]
```diff
@@ -0,0 +1,577 @@
...
+"""
+
+df["label"].value_counts()
+
+"""
+To account for that we will go for a stratified split. 
+"""
+
+# 10% for test
+train_df, test_df = train_test_split(
```
Comment (fchollet) [inline examples/nlp/multimodal_entailment.py:173]: Don't use sklearn for this, just do it by hand, it will be more explicit

File: examples/nlp/multimodal_entailment.py [hunk 5]
```diff
@@ -0,0 +1,577 @@
...
+
+"""
+Our text preprocessing code mostly comes from
+[this tutorial](https://www.tensorflow.org/text/tutorials/bert_glue).
+You are highly encouraged to check out this tutorial to know more about the input
+preprocessing.
+"""
+
+
+def make_bert_preprocess_model(sentence_features, seq_length=128):
```
Comment (fchollet) [inline examples/nlp/multimodal_entailment.py:213]: preprocessing

File: examples/nlp/multimodal_entailment.py [hunk 6]
```diff
@@ -0,0 +1,577 @@
...
+        image = tf.image.decode_png(image, 3)
+    image = tf.image.resize(image, resize)
+    return image
+
+
+def preprocess_text(text_1, text_2):
+    text_1 = tf.convert_to_tensor([text_1])
+    text_2 = tf.convert_to_tensor([text_2])
+    output = bert_preprocess_model([text_1, text_2])
+    output = {feature: tf.squeeze(output[feature]) for feature in bert_input_features}
```
Comment (fchollet) [inline examples/nlp/multimodal_entailment.py:322]: Why the squeeze?

File: examples/nlp/multimodal_entailment.py [hunk 7]
```diff
@@ -0,0 +1,577 @@
...
+
+"""
+### Preprocessing utilities
+"""
+
+resize = (128, 128)
+bert_input_features = ["input_word_ids", "input_type_ids", "input_mask"]
+
+
+def read_resize(image_path):
```
Comment (fchollet) [inline examples/nlp/multimodal_entailment.py:306]: preprocess_image?

File: examples/nlp/multimodal_entailment.py [hunk 8]
```diff
@@ -0,0 +1,577 @@
...
+
+def preprocess_text(text_1, text_2):
+    text_1 = tf.convert_to_tensor([text_1])
+    text_2 = tf.convert_to_tensor([text_2])
+    output = bert_preprocess_model([text_1, text_2])
+    output = {feature: tf.squeeze(output[feature]) for feature in bert_input_features}
+    return output
+
+
+def preprocess(sample):
```
Comment (fchollet) [inline examples/nlp/multimodal_entailment.py:326]: preprocess_text_and_images?

File: examples/nlp/multimodal_entailment.py (hunk 1 above)
Comment (sayakpaul) [reply examples/nlp/multimodal_entailment.py:499]: Nothing particular. I just wanted to verify their implementation correctness w.r.t the ideas.

File: examples/nlp/multimodal_entailment.py [hunk 9]
```diff
@@ -0,0 +1,577 @@
...
+
+batch_size = 32
+auto = tf.data.AUTOTUNE
+
+
+def prepare_dataset(dataframe, training=True):
+    ds = dataframe_to_dataset(dataframe)
+    if training:
+        ds = ds.shuffle(len(train_df))
+    ds = ds.map(lambda x, y: (preprocess(x), y))
```
Comment (fchollet) [inline examples/nlp/multimodal_entailment.py:345]: You should add caching to avoid doing the same preprocessing multiple times. You'll already buffering the data in memory (since you do a global shuffle) so this is free. `.cache()`

**Review by fchollet — COMMENTED** [review]
Thanks for the PR! Super cool example as usual :)

File: examples/nlp/multimodal_entailment.py (hunk 4 above)
Comment (sayakpaul) [reply examples/nlp/multimodal_entailment.py:173]: I wanted to but since we are doing a stratified split I went with `sklearn`.

File: examples/nlp/multimodal_entailment.py (hunk 6 above)
Comment (sayakpaul) [reply examples/nlp/multimodal_entailment.py:322]: It is because the preprocessor adds a batch dimension which wouldn't have been a problem if we had applied the preprocessor on batches. For a single sample, the expected dimension is `(128, )` with the squeeze included. 

Since we are processing the images and the texts simultaneously I tried to keep their preprocessing steps as coherent as possible. To this end, I got confused to apply the following logic to batches:

```python
def read_resize(image_path):
    extension = tf.strings.split(image_path)[-1]

    image = tf.io.read_file(image_path)
    if extension == b"jpg":
        image = tf.image.decode_jpeg(image, 3)
    else:
        image = tf.image.decode_png(image, 3)
    image = tf.image.resize(image, resize)
    return image
```

**Comment by sayakpaul** [comment]
@fchollet addressed most of your comments and also added detailed notes about modality dropout and cross-attention.

File: examples/nlp/multimodal_entailment.py (hunk 6 above)
Comment (fchollet) [reply examples/nlp/multimodal_entailment.py:327]: Do you still get this with TF nightly?

**Review by fchollet — APPROVED** [review]
Awesome! I applied copyedit changes. Please pull them and add the generated files.

**Comment by sayakpaul** [comment]
@fchollet done. Thank you.

**Comment by fchollet** [comment]
Thank you for the great contribution 👍

**Comment by fchollet** [comment]
Also, fun fact: this is the 100th code example on keras.io!

**Comment by sayakpaul** [comment]
Glad to have hit that century. Here's to 500 more. 

Immensely thankful for all that you do to uplift the community.
